corehq/motech/dhis2/finders.py

```diff
@@ -0,0 +1,99 @@
+from collections import namedtuple
+from functools import partial
+
+from memoized import memoized_property
+
+from corehq.motech.dhis2.const import DHIS2_API_VERSION
+from corehq.motech.finders import MATCH_FUNCTIONS, MATCH_TYPE_EXACT
+
+CandidateScore = namedtuple('CandidateScore', 'candidate score')
+
+
+class TrackedEntityInstanceFinder:
+
+    def __init__(self, requests, case_config):
+        self.requests = requests
+        self.case_config = case_config
+
+    @property
+    def property_weights(self):
+        return self.case_config.finder_config.property_weights
+
+    @property
+    def confidence_margin(self):
+        return self.case_config.finder_config.confidence_margin
+
+    @memoized_property
+    def attr_type_id_value_source_by_case_property(self):
+        return {
+            value_source["case_property"]: (attr_type_id, value_source)
+            for attr_type_id, value_source in self.case_config.attributes.items()
+        }
+
+    def find_tracked_entity_instances(self, case_trigger_info):
+        """
+        Search DHIS2 for potential matches of the CommCare case. Score
+        search results and keep those with score > 1. If more than one
+        result has a score > 1, select the best candidate if it exceeds
+        a confidence margin. Otherwise return all results with score > 1.
+        """
+        results = self.fetch_query_results(case_trigger_info)
+        candidate_scores = []
+        for instance in results:
+            score = self.get_score(instance, case_trigger_info)
+            if score >= 1:
+                candidate_scores.append(CandidateScore(instance, score))
+
+        if len(candidate_scores) > 1:
+            candidate_scores = sorted(candidate_scores, key=lambda cs: cs.score, reverse=True)
+            if candidate_scores[0].score / candidate_scores[1].score > 1 + self.confidence_margin:
```

Comment:
This appears to consider only the first two items in `candidate_scores`. Is it intentional that all other items are ignored? Maybe it is not possible for there to be more than two?

Reply:
They are sorted by score. We only care about the candidate with the highest score, and how much greater their score is from the score of the runner up.

e.g. If we are trying to match a case with a tracked entity instance, and we know the case's first name, surname, date of birth and social security number. Getting just the first name and surname right is enough to score 1.0 (not a great idea, but bear with me). DHIS2 has a Jane Doe who is 10 years older and has the wrong SSN; a Jane Doe with the same birthday but no SSN, and a Jane Doe where all 4 properties match.

We don't need to consider that the runner up beat the third-best candidate by the weight associated with date of birth. We only care about the difference in score between the best candidate, and the runner up, because that is what determines how confident we are that the best candidate is the correct candidate.

If you think my reasoning about this is wrong, please let me know. False positives / type I errors are very bad, because they result in CommCare/MOTECH overwriting valid data (usually patient data) with the wrong data! So I want to be sure that we are doing the best possible job here.

Reply:
I can spot one way in which we could improve this algorithm though. Currently we don't differentiate between tracked entity attributes with no value and tracked entity attributes with an incorrect value. If we had a way of weighting incorrect values lower than unknown values, we could have more accurate overall score.